### PR TITLE
[to 48] Add kv_api and update kvproto and related dependences

### LINKED ISSUE #48: TiKV CDC Development Tasks
# TiKV CDC Development Tasks

Related issue: https://github.com/tikv/tikv/issues/11745
Related RFC: https://github.com/tikv/rfcs/pull/86

# Tasks
- [x] Core data structure in ETCD
  - [x] raw changefeed
  - [x] raw task (a part of a changefeed, for one processor)
  - [x] ETCD path planning
- [x] Owner
  - [x] New `changefeed` for RawKV
  - [x] Delete table `changefeed`
- [ ] Scheduler
  - [ ] New scheduler for `Key Span`
  - [x] Delete table scheduler
- [x] Capture
  - [x] Support rawkv `changefeed`
  - [x] Delete support of table `changefeed`
- [x] Processor
  - [x] New processor for Key Span
- [x] Sink
  - [x] TiKV Sink
- [x] Command line
  - [x] cdc cli
- [ ] Others
  - [ ] Upgrade dependencies
- [ ] TiUP support
- [ ] Metrics
  - [ ] New Grafana board for TiKV-CDC
  - [ ] Determine metrics (ref to TiCDC metrics)
- [ ] Tests
  - [ ] Integration tests
  - [ ] E2E tests

# Development Guide
1. Choose a task you are interested in.
2. File a ISSUE for the task.
3. Fix the ISSUE by PR (also refer to CONTRIBUTING guide).
4. Get the PR merged.

### PR DESCRIPTION
**Pull request by zeminzhou**
### What problem does this PR solve?

Add kv_api and update kvproto and related dependences

Issue Number: close #48

Problem Description: **TBD**

### What is changed and how does it work?
1. Update kvproto and add kv_api for ChangeDataRequest;
2. Add "sink key with ttl into downstreams";
3. Update related dependences(In particular, modify go.etcd.io/etcd path);

### Code changes

- Has exported function/method change
- Has exported variable/fields change
- Has interface methods change
- Has persistent data change
- No code

### Check List for Tests

This PR has been tested by at least one of the following methods:
- Unit test

### CODE REVIEW

File: cdc/cdc/sink/tikv.go [hunk 1]
```diff
@@ -204,12 +206,36 @@ type innerBatch struct {
...
-	byteSize int64
+	Batches        map[model.OpType]*innerBatch
+	BatchesWithTTL map[model.OpType]*innerBatch
+	count          int
+	byteSize       int64
+}
+
+func NewTiKVBatcher() *tikvBatcher {
+	tikvBatcher := &tikvBatcher{}
+	tikvBatcher.Batches[model.OpTypePut] = &innerBatch{
```
Comment (pingyu) [inline cdc/cdc/sink/tikv.go:223]: `TTL == 0` is treated as never expired. So we can put `Batches` & `BatchesWithTTL` together.

File: cdc/cdc/sink/tikv.go [hunk 2]
```diff
@@ -221,32 +247,61 @@ func (b *tikvBatcher) ByteSize() int64 {
...
+			return
 		}
+	} else {
+		batcher, ok = b.BatchesWithTTL[entry.OpType]
+		if !ok {
+			log.Info("Unknow operate type, only support put", zap.Int("OpType", int(entry.OpType)))
+			return
+		}
+		// Maybe lose precision
+		batcher.TTL = append(batcher.TTL, entry.ExpiredTs-uint64(time.Now().Unix()))
```
Comment (pingyu) [inline cdc/cdc/sink/tikv.go:267]: Cache `time.Now()`.

File: cdc/cdc/sink/tikv.go [hunk 3]
```diff
@@ -221,32 +247,61 @@ func (b *tikvBatcher) ByteSize() int64 {
...
 	}
-	b.Batches[opType].Keys = append(b.Batches[opType].Keys, entry.Key[1:])
-	b.Batches[opType].Values = append(b.Batches[opType].Values, entry.Value[1:])
+
+	// ignore prefix 'r'
+	key, value := entry.Key[1:], entry.Value
```
Comment (pingyu) [inline cdc/cdc/sink/tikv.go:271]: No need to ignore `r`. `client-go` will deal with the prefix.

File: cdc/cdc/sink/tikv.go [hunk 4]
```diff
@@ -221,32 +247,61 @@ func (b *tikvBatcher) ByteSize() int64 {
...
+	batcher.count += 1
 
 	b.count += 1
-	b.byteSize += int64(len(entry.Key) + len(entry.Value))
+	b.byteSize += int64(len(key) + len(value))
 }
 
 func (b *tikvBatcher) Reset() {
-	b.Batches = map[model.OpType]*innerBatch{}
+	b.Batches[model.OpTypePut] = &innerBatch{
```
Comment (pingyu) [inline cdc/cdc/sink/tikv.go:281]: Just set `Keys` to `Keys[:0]` to reduce memory allocations.

File: cdc/cdc/sink/tikv.go [hunk 5]
```diff
@@ -204,12 +206,36 @@ type innerBatch struct {
...
-	Batches  map[model.OpType]*innerBatch
-	count    int
-	byteSize int64
+	Batches        map[model.OpType]*innerBatch
+	BatchesWithTTL map[model.OpType]*innerBatch
+	count          int
+	byteSize       int64
+}
+
+func NewTiKVBatcher() *tikvBatcher {
```
Comment (pingyu) [inline cdc/cdc/sink/tikv.go:221]: TODO: unit tests.

File: cdc/cdc/sink/tikv.go (hunk 1 above)
Comment (pingyu) [reply cdc/cdc/sink/tikv.go:226]: TODO: fix in next PR.

**Comment by pingyu** [comment]
@zeminzhou @haojinming PTAL

TODO in next PR:
1. Fix comment in `sink/tikv.go`
2. Add unit tests for `sink/tikv.go`
3. Remove `pkg/p2p`

**Review by haojinming — APPROVED** [review]
LGTM~

**Comment by zeminzhou** [comment]
LGTM~

**Review by pingyu — APPROVED** [review]
LGTM

**Pull request merged**